server/src/main/java/io/druid/segment/realtime/firehose/HttpFirehoseFactory.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.segment.realtime.firehose;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import io.druid.data.input.impl.PrefetchableTextFilesFirehoseFactory;
+import io.druid.java.util.common.CompressionUtils;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.util.Collection;
+import java.util.List;
+
+public class HttpFirehoseFactory extends PrefetchableTextFilesFirehoseFactory<URI>
+{
+  private final List<URI> uris;
+
+  @JsonCreator
+  public HttpFirehoseFactory(
+      @JsonProperty("uris") List<URI> uris,
+      @JsonProperty("maxCacheCapacityBytes") Long maxCacheCapacityBytes,
+      @JsonProperty("maxFetchCapacityBytes") Long maxFetchCapacityBytes,
+      @JsonProperty("prefetchTriggerBytes") Long prefetchTriggerBytes,
+      @JsonProperty("fetchTimeout") Long fetchTimeout,
+      @JsonProperty("maxFetchRetry") Integer maxFetchRetry
+  )
+  {
+    super(maxCacheCapacityBytes, maxFetchCapacityBytes, prefetchTriggerBytes, fetchTimeout, maxFetchRetry);
+    this.uris = uris;
+  }
+
+  @JsonProperty
+  public List<URI> getUris()
+  {
+    return uris;
+  }
+
+  @Override
+  protected Collection<URI> initObjects()
+  {
+    return uris;
+  }
+
+  @Override
+  protected InputStream openObjectStream(URI object) throws IOException
+  {
+    return object.toURL().openConnection().getInputStream();
```

Comment:
I think it should be fine. 

1) If prefetch is enabled, 
 1.1) ```PrefetchableTextFilesFirehose``` can fetch http objects in background. When ```next()``` is called, it first checks that there are already fetched files in local storage. If exists, it simply chooses a fetched file and returns a ```LineIterator``` for that file.
 1.2) If there is no fetched files in local storage but some objects are still remained to be read, ```PrefetchableTextFilesFirehose``` fetches one of objects in background immediately. If an IOException occurs while downloading the object, it retries up to the maximum retry count. Finally, ```PrefetchableTextFilesFirehose``` returns a ```LineIterator``` only when the download operation is successfully finished.
2) If prefetch is disabled, ```PrefetchableTextFilesFirehose``` returns a LineIterator which directly reads the stream opened by ```openObjectStream()```. If there is an IOException, it will throw it and the read will fail.

Does it make sense?